File: cpp_advance/oops/inheritance.cpp
// Single inh.

#include <iostream>
using namespace std;

class A{
    protected:
    void a(){
        cout<<"i am from class A."<<endl;
    }
};

class B:public A{
    public:
    void b(){
        a();
        cout<<"i am from class B."<<endl;
    }
    
    
    
};

int main() {
    B obj;
    obj.b();

    return 0;
}

// single inh.

#include <iostream>
using namespace std;

class A{
    public:
    void a(){
        cout<<"i am from class A."<<endl;
    }
};

class B:public A{
    public:
    void b(){
        cout<<"i am from class B."<<endl;
    }
    
    
    
};

int main() {
    B obj;
    obj.a();
    obj.b();

    return 0;
}


// multi-level inheritance
#include <iostream>
using namespace std;

class A{
    public:
    void a(){
        cout<<"i am from class A."<<endl;
    }
};

class B:public A{
    public:
    void b(){
        cout<<"i am from class B."<<endl;
    }
};

class C:public B{
    public:
    void c(){
        cout<<"i am from class C."<<endl;
    }
};

int main() {
    C obj;
    obj.a();
    obj.b();
    obj.c();

    return 0;
}

// multiple inh.
#include <iostream>
using namespace std;

class A{
    public:
    void a(){
        cout<<"i am from class A."<<endl;
    }
};

class B{
    public:
    void b(){
        cout<<"i am from class B."<<endl;
    }
};

class C:public A,public B{
    public:
    void c(){
        cout<<"i am from class C."<<endl;
    }
};

int main() {
    C obj;
    obj.a();
    obj.b();
    obj.c();

    return 0;
}


#include <iostream>
using namespace std;

class Shape{
  public:
  void shape(){
      cout<<"This is a Shape class"<<endl;
  }
};

class Shape2d:public Shape{
  public:
  void shape2d(){
      cout<<"This is a Shape2d class"<<endl;
  }
};

class Circle:public Shape2d{
    public:
    void circle(){
        cout<<"This is a circle class"<<endl;
    }
};
class Square:public Shape2d{
    public:
    void square(){
        cout<<"This is a Square class"<<endl;
    }
};

class Shape3d:public Shape{
  public:
  void shape3d(){
      cout<<"This is a Shape3d class"<<endl;
  }
};

class Cube:public Shape3d{
    public:
    void cube(){
        cout<<"This is a cube class"<<endl;
    }
};


int main() {
    Square obj;
    obj.square();
    obj.shape2d();
    obj.shape();
    

    return 0;
}

File: python/advance/oops/inheritance.py
# class A:
#     def a(self):
#         print("Class A")

# class B(A):
#     def b(self):
#         print("Class B")
# obj = B()
# # print(dir(obj))
# obj.a()
# obj.b()


# class A:
#     def a(self):
#         print("Class A")

# class B(A):
#     def b(self):
#         print("Class B")

# class C(B):
#     def c(self):
#         print("Class C")

# obj = C()
# obj.a()
# obj.b()
# obj.c()


# class A:
#     def a(self):
#         print("Class A")

# class B:
#     def b(self):
#         print("Class B")

# class C(A,B):
#     def c(self):
#         print("Class C")

# obj = C()
# obj.a()
# obj.b()
# obj.c()

class Shape:
    def shape(self):
        print("Shape")

class Shape2d(Shape):
    def shape2d(self):
        print("2D Shape")

class Circle(Shape2d):
    def circle(self):
        print("Circle")

class Rectangle(Shape2d):
    def rectangle(self):
        print("Rectangle")

class Shape3d(Shape):
    def shape3d(self):
        print("3D Shape")

class Cube(Shape3d):
    def cube(self):
        print("Cube")

# obj = Cube()

# obj.shape()

# # obj.shape2d()

# obj.shape3d()

# obj.cube()

# obj = Circle()

# obj.shape()

# obj.shape2d()

# obj.circle()
